pkg/report/title_stat.go

```diff
@@ -86,3 +78,67 @@ func (ts *titleStat) visit(cb func(int, ...string), titles ...string) {
 		ts.Nodes[title].visit(cb, append(titles, title)...)
 	}
 }
+
+func (ts *TitleStat) ToBytes() ([]byte, error) {
+	return json.MarshalIndent(ts, "", "\t")
+}
+
+type TitleFreqRank struct {
+	Title string
+	Count int
+	Total int
+	Rank  int
+}
+
+func (ts *TitleStat) Explain() []*TitleFreqRank {
+	titleCount := map[string]int{}
+
+	ts.visit(func(count int, titles ...string) {
+		uniq := map[string]bool{}
+		for _, title := range titles {
+			uniq[title] = true
+		}
+		for title := range uniq {
+			titleCount[title] += count
+		}
+	})
+	var res []*TitleFreqRank
+	for title, count := range titleCount {
+		res = append(res, &TitleFreqRank{
+			Title: title,
+			Count: count,
+			Total: ts.Count,
+			Rank:  TitlesToImpact(title),
+		})
+	}
+	sort.Slice(res, func(l, r int) bool {
+		if res[l].Rank != res[r].Rank {
+			return res[l].Rank > res[r].Rank
+		}
+		lTitle, rTitle := res[l].Title, res[r].Title
+		if titleCount[lTitle] != titleCount[rTitle] {
+			return titleCount[lTitle] > titleCount[rTitle]
+		}
+		return lTitle < rTitle
+	})
+	return res
+}
+
+// HigherRankTooltip generates a prioritized list of titles with a rank higher than firstTitle.
+func HigherRankTooltip(firstTitle string, titlesInfo []*TitleFreqRank) string {
+	baseRank := TitlesToImpact(firstTitle)
+	res := ""
+	for _, ti := range titlesInfo {
+		if ti.Rank <= baseRank {
+			continue
+		}
+		res += fmt.Sprintf("[rank %2v, freq %5.1f%%] %s\n",
+			ti.Rank,
+			100*float32(ti.Count)/float32(ti.Total),
+			ti.Title)
+	}
+	if res != "" {
+		return fmt.Sprintf("[rank %2v,  originally] %s\n%s", baseRank, firstTitle, res)
```

Comment:
Nit: are two whitespaces here intended?

Reply:
Yes. It aligns the `[.*] ` prefix in all the lines.